src/java/org/apache/cassandra/service/reads/DigestResolver.java

```diff
@@ -109,7 +107,8 @@ public boolean responsesMatch()
         // validate digests against each other; return false immediately on mismatch.
         ByteBuffer digest = null;
         Collection<Message<ReadResponse>> snapshot = responses.snapshot();
-        if (snapshot.size() <= 1)
+        assert snapshot.size() > 0 : "Attempted response match comparison while no responses have been received.";
+        if (snapshot.size() == 1)
```

Comment:
@aholmberg @belliottsmith I think I just hit this in one of my clusters, and I'm trying to recall the reason(s) this assertion was strengthened (...or the reason why it was more permissive beforehand).

```
message="Unexpected exception during request; channel = [id: 0xabc818ff, L:/10.38.178.28:9042 - R:/17.58.81.38:55990]"
exception="java.lang.AssertionError: Attempted response match comparison while no responses have been received.
	at org.apache.cassandra.service.reads.DigestResolver.responsesMatch(DigestResolver.java:110)
	at org.apache.cassandra.service.reads.AbstractReadExecutor.awaitResponses(AbstractReadExecutor.java:393)
	at org.apache.cassandra.service.StorageProxy.fetchRows(StorageProxy.java:2150)
	at org.apache.cassandra.service.StorageProxy.readRegular(StorageProxy.java:1979)
	at org.apache.cassandra.service.StorageProxy.read(StorageProxy.java:1882)
	at org.apache.cassandra.db.SinglePartitionReadCommand$Group.execute(SinglePartitionReadCommand.java:1121)
	at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:296)
	at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:248)
	at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:90)
```

Reply:
I guess previously it was strictly correct - they all matched if there were none.

Reply:
It looks like our `DigestResolver` has a `dataReponse` yet somehow does not have a single element in `responses`. How is that possible, given the read callback should assign both of them in `DigestResolver#preprocess()`? Two things come to mind here:

1.) The JavaDoc for `Accumulator#add()`

```
/**
 * Adds an item to the collection.
 *
 * Note it is not guaranteed to be visible on exiting the method, if another add was happening concurrently;
 * it will be visible once all concurrent adds (which are non-blocking) complete, but it is not guaranteed
 * that any size change occurs during the execution of any specific call.
 *
 * @param item add to collection
 */
```

Is it possible that a lagging response is being added concurrently and hence the size is not updated when we snapshot the `Accumulator`? If that were the case, it seems like relaxing the assertion would allow the query to complete w/ a data response in hand. It's not clear to me that is safe though, as opposed to simply waiting for the `Accumulator` state to settle by the time we signal.

2.) Short read protection

The only other place in the code I can see `responses` being cleared is in the pre-fetch callback for `ShortReadProtection`, but that shouldn't happen until responses have already been consumed and more rows are requested. CC @adelapena

Reply:
(1) is a bug Alex Petrov was working on recently, and is definitely a real issue that could trigger this.

Reply:
I've always kind of wondered why `ReadCallback#onResponse()` doesn't ignore responses that arrive after we've already signaled.